### ISSUE TITLE
TRAVIS_BRANCH env var is lost for tagged commits

### ISSUE THREAD

**Issue by liddellj** [travis-ci/travis-ci#4745]
When building a tagged commit on Travis, the original value of `TRAVIS_BRANCH` is lost, and is instead replaced with the tag name. This has been raised before (#1670, #1674, #1675) and as a result a change was introduced to separately expose the tag name through the `TRAVIS_TAG` env var. While the introduction of `TRAVIS_TAG` is useful, this does not fully address the issue - the value of `TRAVIS_BRANCH` env var is still lost. For one thing - this is not clearly spelled out in the documentation for `TRAVIS_BRANCH`:

> For builds not triggered by a pull request this is the name of the branch currently being built; whereas for builds triggered by a pull request this is the name of the branch targeted by the pull request (in many cases this will be master).

Based on the documentation (and the name of the env var) - I assumed that `TRAVIS_BRANCH` should *always* contain the current git branch name, and `TRAVIS_TAG` should contain the current tag name, if any (which it does). The fact that `TRAVIS_BRANCH` is overloaded to contain the tag name is non-obvious and appears to be causing a lot of confusion.

I realise that changing this behaviour now would potentially introduce a breaking change for customers who have come to rely on the existing behaviour. Therefore I can only see 2 courses of action:
* Introduce a separate env var entirely, which always stores the 'true' branch name, e.g. `GIT_BRANCH`.
* For the sake of keeping the env vars clean and consistent - introduce a breaking change - and make sure the existing `TRAVIS_BRANCH` var is not overwritten with the tag name for tagged builds. In this case you could also introduce `TRAVIS_BRANCH_OR_TAG` to allow customers who are currently relying on the existing behaviour of `TRAVIS_BRANCH` to easily migrate.

My personal scenario here is that I would like to trigger a deploy only for tagged builds, and only on the master branch. Currently this is not possible without some manual scripting because using the `on: tags` condition causes the `branch` condition to be ignored (which I believe would not be the case, were the behaviour of the `TRAVIS_BRANCH` env var to be 'corrected'). Ideally I would like to be able to do:
```
deploy:
  on:
    tags: true
    branch: master
```

But this is currently not supported.

**Comment by liddellj** [comment]
This issue is compounded by the fact that my efforts to use alternate methods to determine the git branch have failed.

> $(git branch | sed -n -e 's/^\* \(.*\)/\1/p')

Comes up empty (I believe because of the way git clone is performed).

> $(git symbolic-ref HEAD --short)

Results in an error: `fatal: ref HEAD is not a symbolic ref`.

So as far as I can tell, Travis makes it *impossible* to determine the branch name for tagged builds. :disappointed:

Check out https://travis-ci.org/liddellj/node-aws-travis-playground/builds/78551189 for an example build that tests these techniques.

**Comment by liddellj** [comment]
For anyone encountering the same issue, the best solution I could come up with was to firstly restrict my local release scripts from being executed on anything but the master branch. Here's an excerpt from my `Makefile`:
```
pre-release: test
	if [ "$(GIT_BRANCH)" != "master" ]; then \
		echo "Current branch is $(GIT_BRANCH). Versioning must be performed on master."; \
		exit 1; \
	fi;

release-patch: pre-release
	npm version patch
	git push --follow-tags

release-minor: pre-release
	npm version minor
	git push --follow-tags

release-major: pre-release
	npm version major
	git push --follow-tags
```
Combined with an extra condition in `.travis.yml` to restrict deployments to tags which look like version tags:
```
deploy:
  on:
    tags: true
    condition: "$TRAVIS_TAG =~ ^v[0-9].[0-9].[0-9]$"
```

This means that *only* tags of the form _v1.2.3_ will trigger a deployment, and the release script can only be executed on master. However, it does not prevent someone from directly executing `git tag` on a non-master branch which happens to be of the form _v1.2.3_, but I think this is unlikely.

**Comment by BanzaiMan** [comment]
I believe this is a duplicate of #4643. Please provide additional details if you think otherwise. Thanks.

**Comment by liddellj** [comment]
Hmm. I guess #4643 sort of describes the same issue - but a much more specific case. As long as the underlying issue is understood to be 'we need a way to determine the branch regardless of the presence of a tag' then I'm happy to close this.

**Comment by BanzaiMan** [comment]
@liddellj Thanks for the feedback. I'm reopening this as a representative case.

**Comment by eteubert** [comment]
In https://github.com/travis-ci/travis-ci/issues/4802#issuecomment-140379320 I explored / tried to explain why there is no branch information available on tag-based clones.

The only workaround I can think is cloning differently: do a clone without specifying `--branch`, then switch to the tag. I don't know the best way to change the clone behaviour though. I opened an SO question for it: http://stackoverflow.com/questions/32580821/how-can-i-customize-override-the-git-clone-step-in-travis-ci

Ideally this could be solved directly in travis though :)

**Comment by mdumrauf** [comment]
I am in the need of knowing the current branch too. +1. Please consider fixing this or adding a new environment variable.

**Comment by lexman** [comment]
Hello,
any news on this issue ?

**Comment by danepowell** [comment]
As you can see there's quite a number of issues piling up in the community all referencing this one... getting this fixed would make a lot of people very happy.

**Comment by BanzaiMan** [comment]
Tags on git are associated with a commit object, not a branch. You can tag a commit without any regards to the branch. You can tag a detached commit.

GitHub's API reflects this: https://developer.github.com/v3/git/tags/ When we are notified of the new tag, the payload does not contain any information about the branch at all.

**Comment by danepowell** [comment]
I definitely get that the branch name may not be available in all cases. But at the very least the documentation should be updated to reflect that fact, since right now it indicates that the TRAVIS_BRANCH _is_ available even for tagged commits.

**Comment by brightrnservice** [comment]
i tried using `git branch -a --contains $COMMITID` but the branch name that should appear according the commit id is not shown in tagged build

**Comment by timusus** [comment]
@BanzaiMan right, but the travis build gets its configuration file from `travis.yml` on some branch, right? 

When a tag triggers a build, presumably all branches containing the tagged commit and a `travis.yml` file will be built? And if that is the case, then shouldn't each one of those builds know something about which branch it's being built from?

**Comment by BanzaiMan** [comment]
@timusus 
> right, but the travis build gets its configuration file from travis.yml on some branch, right?

It does not.

```sh-session
$ git commit --allow-empty -m "Empty commit"
[default 061c35c] Empty commit
$ git checkout $(git rev-list HEAD -1)
Note: checking out '061c35c5cda1dcd2fc3b744587272e5a6f4e58c8'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

HEAD is now at 061c35c... Empty commit
$ git tag detached-commit -f
Updated tag 'detached-commit' (was e4b17e7)
$ git push origin --tags -f
Counting objects: 1, done.
Writing objects: 100% (1/1), 187 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
To git@github.com:BanzaiMan/travis_production_test.git
 + e4b17e7...061c35c detached-commit -> detached-commit (forced update)
```

https://travis-ci.org/BanzaiMan/travis_production_test/builds/210183248 is the resulting build, and the commit https://github.com/BanzaiMan/travis_production_test/commit/061c35c5cda1dcd2fc3b744587272e5a6f4e58c8 does not have any branch information, but a tag.

**Comment by LorensK** [comment]
If it is not possible to know the branch a tagged commit is coming from, then **PLEASE** pretty please at least update the [documentation of variable TRAVIS_BRANCH](https://docs.travis-ci.com/user/environment-variables/) to reflect this.

**Comment by BanzaiMan** [comment]
The documentation now reflects that `TRAVIS_BRANCH` is set to the tag name when the build was initiated by a tag.

However, it seems also reasonable to me that `TRAVIS_BRANCH` should be empty for tag builds. I do wonder if this change would break existing builds, however.

**Comment by altaurog** [comment]
It is possible to directly query the branch name(s) from the source repo.  Here is [the answer](https://stackoverflow.com/a/47393948/519015) I gave to the [SO question](https://stackoverflow.com/q/32580821/519015) referenced above:
```
env:
    global:
        # get all the branches referencing this commit
        - REAL_BRANCH=$(git ls-remote origin | sed -n "\|$TRAVIS_COMMIT\s\+refs/heads/|{s///p}")

        # or check if we are on a particular branch:
        - IS_RELEASE=$(git ls-remote origin | grep "$TRAVIS_COMMIT\s\+refs/heads/release$"
```

(This is relevant also for issues #1670, #4645, #7326, and travis-ci/docs-travis-ci-com#971)
